### ISSUE TITLE
Milestone 2: Implement the mesh refinement criterion

### ISSUE THREAD

**Issue by bangerth** [bangerth/helmholtz#10]
As stated in the readme.md, we want to do this:
```
On the other hand, if the number of mesh refinement steps is negative,
then the mesh is adapted using the following algorithm. For a given
wave speed, we can compute the wavelength of oscillations for a
given frequency _w_ via
```
  lambda = c/f
         = c/(w/(2pi))
```
If `lambda` is larger than the diameter of the domain, then we replace
it by the diameter of the domain since that is the largest wavelength
at which the solution can vary.
We can then determine a frequency-adapted mesh size in the following
way: We want that there are at least _N_ grid points per wave
length. Since there are as many grid points per cell as the polynomial
degree _p_, this equations to requiring that the mesh size _Delta x_
satisfies _Delta x / p <= lambda/N_
or equivalently: _Delta x <= lambda/N * p_. The `Mesh refinement steps`
parameter is then interpreted as _N_ if it is negative.
```

In the code, we just need to compute lambda correctly -- right now it is simply set to 1000.

**Comment by bangerth** [comment]
From Jason:

> The solution will be of the form:
>  
> exp(-j*k*x)
>  
> where k= w/c.
>  
> if we break this up into it's real and imaginary parts  k = kr - j*ki :
>  
> exp(-j*kr*x)*exp(-ki*x) 
>  
> we'll get a rapid transition when
>  
> kr*x > 2*pi/10  (1/10th of a wavelength)
>  
> or when 
>  
> exp(-ki*x) < 0.9 ???    or   x > -loge(0.9)/ki  since loge(0.9)~= 1/10 we have r   x > 1/(10*ki)  ???
>  
> so we want the cell size x to be smaller than
>  
> 2*pi/(10*kr)   and   1/(10*ki)  ???
>  
> which ever is smaller???   Since we expect ki << kr, usually the 2*pi*(10*kr) factor will control the cell size, but if there's a lot of damping, the ki term could.

That's with ten points per wave length. If we did _N_ points per wave length, then we need 
```
  h <= min (2*pi/(N*kr),  1/(N*ki) )
```

**Comment by bangerth** [comment]
Implemented in 9fb20f2df7eec9ea1a117b619b47a6bfb3b4ae4c.

**Comment by bangerth** [comment]
Documented in bcdba18e001dd54bb5b5f79f39b744cad2d9a5af.
